### PR TITLE
WIP: Docker / Docker-compose for deploying intercity-next

### PR DESCRIPTION
**Pull request by ariejan**
- [x] Containerize intercity-next with postgres and redis
- [x] Add nginx-proxy and expose ports 80/443
- [x] Add Letsencrypt support
- [x] Add setup script to kick-off deployment

Currently known issues

- [x] Stuck on spinner when adding a new server (validating ssh) 
- [x] No services/plugins seem to have been installed on a new server (no pgsql, redis, etc.)
- [ ] Ugly 502 Bad Gateway error when booting

### CODE REVIEW

**Comment by ariejan** [comment]
Steps to setup a new Intercity Next machine:
 * Create a server (tested on Ubuntu 16.04)
 * Clone intercity-next
 * Run ./scripts/setup which will setup docker/docker-compose for you and kick off the installation
 * Wait for containers to boot, database to be migrated and SSL certs to be requested, takes a minute or so.
 * Done. 

Known issues:
 * Not sure if it's related to this setup, but adding a new server gets stuck at verifying the SSH key.

File: Procfile [hunk 1]
```diff
@@ -1,2 +1,2 @@
-web: bundle exec rails s
-sidekiq: bundle exec sidekiq
+web: bundle exec puma -t 5:5 -p ${PORT:-3000} -e ${RACK_ENV:-development}
```
Comment (jvanbaarsen) [inline Procfile]: What is the benefit of using puma explicitly over rails s? Rails uses Puma out of the box when it is in the gemfile right? Or is that only in later rails versions?

File: scripts/setup [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+fi
+
+echo "-- Deploying Intercity Next to https://$INTERCITY_FQDN"
+
+# Boot up with docker compose
+docker-compose up -d
+
+echo "-- Waiting for Intercity Next to fully boot up. This may take a few minutes."
+
+until $(curl --output /dev/null --silent --head --fail https://$INTERCITY_FQDN); do
```
Comment (jvanbaarsen) [inline scripts/setup]: ❤️

File: config/initializers/secret_token.rb [hunk 3]
```diff
@@ -1,10 +1,12 @@
...
-
-token = $redis.get("SECRET_KEY_BASE")
-if token.nil?
-  token = SecureRandom.hex(64)
-  $redis.set("SECRET_KEY_BASE", token)
-end
-
-Rails.application.config.secret_token = token
-Rails.application.config.secret_key_base = token
+# require 'securerandom'
```
Comment (jvanbaarsen) [inline config/initializers/secret_token.rb]: What is the reason for commenting this?

**Review by jvanbaarsen — COMMENTED** [review]
I really love this direction! Looks super straight forward. 

How does this affect development for the app? Will that still be the same as it was before?

File: Procfile (hunk 1 above)
Comment (ariejan) [reply Procfile]: I think Rails does that, but I like to be explicit and avoid using Webrick by accident.

File: config/initializers/secret_token.rb (hunk 3 above)
Comment (ariejan) [reply config/initializers/secret_token.rb]: Yeah, about that. `config/application.rb` created a new instance of `Redis` (with default values). But for Docker we need to use `REDIS_URL`, which is not yet available in `application.rb`. 
Also, _using_ Redis in initializers requires you have redis setup when running _any_ rake task (like `assets:precompile` and `db:migrate`). 

I'm not sure what the use-case was for storing the secret token in Redis, I assume keeping it alive between deployments. I've now defined it in `.env`/`env.example` instead. 

If the above is okay with you, the whole `config/initializers/secret_token.rb` can be deleted.

**Comment by ariejan** [comment]
The SSH spinner issues has been resolved by downgrading ruby to 2.4.2. There's some changes in how Ruby handles private keys, and this was the fastest way to work around it. Upgrading to Ruby 2.5.1 will be a bit more involved because of that. 

Next issues: docker/dokku are installed fine, but not the extra services, like postgresql, redis and mariadb.

File: Procfile (hunk 1 above)
Comment (jvanbaarsen) [reply Procfile]: 👍

File: config/initializers/secret_token.rb (hunk 3 above)
Comment (jvanbaarsen) [reply config/initializers/secret_token.rb]: @ariejan It was something I took over from GitLab. Mostly so users don't have to generate their own key. That was something that went wrong quite some times over there. The secret key needs to be a certain length for it to work, and people tend to mess that up. So for me that was the reason to generate it on the fly.

How do you think we can tackle the "human" issue?

Comment (jvanbaarsen) [reply config/initializers/secret_token.rb]: Ah yeah! That sounds good.

Comment (ariejan) [reply config/initializers/secret_token.rb]: I've update setup to ask for all info that's needed to generate a `.env`. `SECRET_KEY_BASE` is generated as well.

**Comment by ariejan** [comment]
@jvanbaarsen is it necessary to run `rake intercity:seed_plugins` when installing or even when booting? Seems to be the reason I don't have any services installed.

**Comment by jvanbaarsen** [comment]
@ariejan Aaah yeah good one! The reason that seed task is in there is that I needed a simple way to add new services without having to login on all instances that we have running to update the available services list.

**Comment by jvanbaarsen** [comment]
@ariejan I'm trying out this PR locally, but I can't seem to get it to work. What domain did you use to test this? Does it have to be a domain that is reachable for letsencrypt? Or can I just use localhost?

**Comment by ariejan** [comment]
I spun up a €3 vps with hetzner and a real fqdn.   
  
  
  

  
  
  
  
  
>

**Comment by jvanbaarsen** [comment]
@ariejan Looks like we can solve the 502 errors with the following snippet in the entrypoint script:

```
echo "-- Waiting for DB to be active"
while ! pg_isready -h db > /dev/null 2> /dev/null; do
  printf "."
  sleep 1
done
```

We would then also need to add `postgresql-client` to the docker file as a dependency to install.

For me this at least solved some of the errors in the logs when the app started earlier then the DB is active.

**Comment by michiels** [comment]
Very intersting stuff @ariejan! I'll also be checking this out. See if it can be combined or run alongside my other PR where I started a docker-compose setup for local development (then you don't need the nginx and letsencrypt proxies and other magic).

File: config/database.yml [hunk 4]
```diff
@@ -14,5 +14,7 @@ test:
 production:
   <<: *default
   database: intercity
-  host: <%= ENV["DB_SOCKET"] %>
+  host: db
```
Comment (michiels) [inline config/database.yml]: Shall we perhaps use ENV vars here too? That allows for running the app with external or offloaded databases too. For example, at Firmhouse I would offload the database to an encrypted-at-rest Database service that's maintained by a 3rd party.

File: scripts/setup [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    printf '.'
+    sleep 5
+done
+
+echo
+echo
+echo "Intercity Next has been started and is ready for departure!"
```
Comment (michiels) [inline scripts/setup]: 🤣

**Comment by michiels** [comment]
Hi @ariejan 

I'm planning on picking up deploying Intercity as Docker (Compose) app again, but you've already done massive important work in this PR Is this PR something you'd like to continue on and ship? Or would you like me to take over?

Thanks for your reply!

Michiel

**Comment by ariejan** [comment]
@michiels due to time constraints I cannot work on this right now, so feel free to continue working on it. Feel free to mention me here if you have any questions.

**Comment by michiels** [comment]
Hi @ariejan Perfect! Thank you for your swift reply. I'll analyze your PR and do some follow up steps :)

**Pull request closed without merging**